sessions/debugging.qmd

```diff
@@ -1,24 +1,192 @@
+    green arrow on the left of the line number.
+
+-   The Console will now start with `Browse[1]>` and will have text like
+    `debug at ...`.
+
+-   There will be new buttons on the top of the Console like "Next",
+    "Continue", and "Stop".
+
+-   The Environment pane will be empty and will say "Traceback".
+
+### Using the debugger
+
+```{=html}
+<!--#TODO:
+Maybe add a more complex example with a less informative error message, similar to the long call stack example from adv-R
+-->
+```
+We might have a function like this, which returns the double of the
+input minus the squareroot of the input:
+
+```{r, eval=FALSE}
+my_function <- function(number) {
+  double_of_number <- number + number
+  number_squareroot <- sqrt(number)
+  return(double_of_number - number_squareroot)
+}
+```
+
+When you use this function, it usually works without a problem, but some
+times it returns `NaN` with a warning. Since this is a small toy
+function with a helpful warning message, you may be able to spot the
+location of the issue right away, but you may not always be so lucky.
+The debugger can always help you, so let's use it to find out where the
+problem arises in the function!
+
+Let's start the debugger by inserting a `browser()` statement inside the
+function:
+
+```{r, eval=FALSE}
+test_debugging <- function(number) {
+  browser()
+  double_of_number <- number + number
+  number_squareroot <- sqrt(number)
+  return(double_of_number - number_squareroot)
+}
+```
+
+And re-run and use the function again, which will start up the debugger
+whenever the function is called. Now we can inspect the number variable
+in the function environment, and we can change it and see what happens
+to the output. It quickly becomes clear which part of the code is
+causing the `NA` outputs and why it is causing them.
+
+## Using stop()
+
+By default, R is quite forgiving about accepting inputs, even those that
+might not be the intended type. Even when it might not make much sense,
+R will try its best to use the inputs and run the function you ask it
+to.
+
+In the above example, we got a warning with an informative message
+related to the `NAs` produced by the `sqroot()` function, but sometimes
+it might be best to return an error and stop the code from executing, or
+to add a specific error message. The `stop()` function allows you to do
+just that, and helps the user start the debugger in the area of the code
+and environment that caused the error. This can be very helpful to
+"future you" or other users, so consider where a `stop()` call can
+protect users from unexpected behavior when your are making custom
+functions.
+
+```{r, eval=FALSE}
+test_debugging <- function(number) {
+  double_of_number <- number + number
+    if(number < 0) {
+    stop("Input number is negative, can't compute square root")
+  }
+  number_squareroot <- sqrt(number)
+  
+  return(double_of_number - number_squareroot)
+}
+```
+
+#### Exercise ideas
```

Comment:
Just some other ideas too:

- Give them some code that has something wrong in it and get them to run the debugger to find the problem
- Get them to find the best place to insert abort and warn functions